lib/tbot/service/service.go

```diff
@@ -0,0 +1,187 @@
+//
+// If the given context is canceled or reaches its deadline, Wait will unblock
+// immediately and return the context's error. This is useful for implementing
+// timeout/fallback behavior.
+//
+// If the service finishes running without becoming Ready (i.e. in one-shot
+// mode), ErrWrongStatus will be returned.
+func (s *Service[HandlerT]) Wait(ctx context.Context) (HandlerT, error) {
+	return s.WaitForStatus(ctx, status.Ready)
+}
+
+// WaitForStatus blocks until the service's status matches one of the given
+// statuses and then returns the handler.
+//
+// If the given context is canceled or reaches its deadline, WaitForStatus will
+// unblock immediately and return the context's error. This is useful for
+// implementing timeout/fallback behavior.
+//
+// If the service finishes running before reaching the desired status (i.e. in
+// one-shot mode), ErrWrongStatus will be returned.
+func (s *Service[HandlerT]) WaitForStatus(ctx context.Context, statuses ...status.Status) (HandlerT, error) {
+	notifCh, close := s.status.ChangeNotifications()
+	defer close()
+
+	if slices.Contains(statuses, s.status.Get()) {
+		return s.handler, nil
+	}
+
+	var zero HandlerT
+	for {
+		var finalValue, ctxDone bool
+		select {
+		case <-notifCh:
+			// New value ready.
+		case <-s.done:
+			finalValue = true
+		case <-ctx.Done():
+			ctxDone = true
+		}
+
+		switch {
+		case slices.Contains(statuses, s.status.Get()):
+			return s.handler, nil
+		case finalValue:
+			return zero, trace.Wrap(ErrWrongStatus)
+		case ctxDone:
+			return zero, trace.Wrap(ctx.Err())
+		default:
+			// Wait for the next update.
+		}
+	}
+}
+
+// These methods implement the InternalService API used by the supervisor.
+func (s *Service[HandlerT]) getName() string                     { return s.name }
+func (s *Service[HandlerT]) setStatus(status status.Status) bool { return s.status.Set(status) }
+func (s *Service[HandlerT]) runHandler(ctx context.Context, runtime *Runtime) error {
+	return s.handler.Run(ctx, runtime)
+}
+func (s *Service[HandlerT]) runOneShotHandler(ctx context.Context) error {
+	var handler any = s.handler
+	if os, ok := handler.(OneShotHandler); ok {
+		return os.OneShot(ctx)
+	}
+	return errNoOneShotHandler
+}
+func (s *Service[HandlerT]) finalize() { close(s.done) }
+func (s *Service[HandlerT]) registerToSupervisor() bool {
+	return s.registered.CompareAndSwap(false, true)
+}
+
+// errNoOneShotHandler is a sentinel error used to tell the supervisor that the
+// service doesn't support one-shot mode.
+var errNoOneShotHandler = errors.New("supervisor: handler does not implement OneShotHandler")
+
+// Runtime allows service handlers to communicate with the supervisor and other
+// services.
+type Runtime struct{ setStatusFn func(status status.Status) }
+
+// SetStatus communicates the service's status with the supervisor and other
+// services.
+func (rt *Runtime) SetStatus(status status.Status) { rt.setStatusFn(status) }
```

Comment:
Do you have plans to extend this? It feels simpler to pass `func(status status.Status)` around. Or even type it to `type StatusSetter func(status status.Status)` or something.

Reply:
Yeah I'm wondering about using it to pass in the logger too, but can't make up my mind on that one 😄